UnityProject/Assets/Scripts/Electricity/PoweredDevices/APC.cs

```diff
@@ -0,0 +1,450 @@
+	/// </summary>
+	private bool RefreshDisplay = false;
+
+	private void StartRefresh()
+	{
+		gameObject.SetActive(true);
+		RefreshDisplay = true;
+		StartCoroutine(Refresh());
+	}
+	public void RefreshOnce()
+	{
+		RefreshDisplay = false;
+		StartCoroutine(Refresh());
+	}
+	private void StopRefresh()
+	{
+		RefreshDisplay = false;
+	}
+	private IEnumerator Refresh()
+	{
+		RefreshDisplayScreen();
+		yield return WaitFor.Seconds(2f);
+		if (RefreshDisplay)
+		{
+			StartCoroutine(Refresh());
+		}
+	}
+
+	//	///// <summary>
+	//	///// Animates the APC screen sprites
+	//	///// </summary>
+	private void RefreshDisplayScreen()
+	{
+		if (++displayIndex >= loadedScreenSprites.Length)
+		{
+			displayIndex = 0;
+		}
+		screenDisplay.sprite = loadedScreenSprites[displayIndex];
+	}
+
+	// -----------------------------------------------------
+	//					CONNECTED LIGHTS AND BATTERY THINGS
+	// -----------------------------------------------------
+	/// <summary>
+	/// The list of emergency lights connected to this APC
+	/// </summary>
+	private List<EmergencyLightAnimator> ConnectedEmergencyLights = new List<EmergencyLightAnimator>();
+
+	/// <summary>
+	/// Devices connected to APC
+	/// </summary>
+	public List<APCPoweredDevice> ConnectedDevices = new List<APCPoweredDevice>();
+
+	// TODO make apcs detect connected department batteries
+	/// <summary>
+	/// List of the department batteries connected to this APC
+	/// </summary>
+	public List<DepartmentBattery> ConnectedDepartmentBatteries = new List<DepartmentBattery>();
+
+	private bool _emergencyState = false;
+	public void OnDespawnServer(DespawnInfo info)
+	{
+		for (int i = ConnectedDevices.Count-1; i >= 0; i--)
+		{
+			ConnectedDevices[i].RemoveFromAPC();
+		}
+	}
+	/// <summary>
+	/// The state of the emergency lights. Calls SetEmergencyLights when changes.
+	/// </summary>
+	private bool EmergencyState
+	{
+		get
+		{
+			return _emergencyState;
+		}
+		set
+		{
+			if (_emergencyState != value)
+			{
+				_emergencyState = value;
+			}
+		}
+	}
+
+	#region Editor
+
+	void OnDrawGizmosSelected()
+	{
+		var sprite = GetComponentInChildren<SpriteRenderer>();
+		if (sprite == null)
+			return;
+
+		//Highlighting all controlled lightSources
+		Gizmos.color = new Color(0.5f, 0.5f, 1, 1);
+		for (int i = 0; i < ConnectedDevices.Count; i++)
+		{
+			var lightSource = ConnectedDevices[i];
+			if(lightSource == null) continue;
+			Gizmos.DrawLine(sprite.transform.position, lightSource.transform.position);
+			Gizmos.DrawSphere(lightSource.transform.position, 0.25f);
+		}
+	}
+
+	#endregion
+
+	//######################################## Multitool interaction ##################################
+	[SerializeField]
+	private MultitoolConnectionType conType = MultitoolConnectionType.APC;
+	public MultitoolConnectionType ConType  => conType;
+
+	[SerializeField]
+	private bool multiMaster = true;
+	public bool MultiMaster  => multiMaster;
+
+	public void AddSlave(object SlaveObject)
```

Comment:
Codacy found an issue: [Add a nested comment explaining why this method is empty, throw a 'NotSupportedException' or complete the implementation.](https://app.codacy.com/gh/unitystation/unitystation/pullRequest?prid=6280689)